File: internal/controller/rules/home/autoaway.go
package home

import (
	"fmt"
	"github.com/clambin/go-common/set"
	"github.com/clambin/tado-exporter/internal/controller/rules"
	"github.com/clambin/tado-exporter/internal/controller/rules/action"
	"github.com/clambin/tado-exporter/internal/controller/rules/configuration"
	"github.com/clambin/tado-exporter/internal/poller"
	"github.com/clambin/tado/v2"
	"log/slog"
	"strings"
	"time"
)

type AutoAwayRule struct {
	usernames set.Set[string]
	delay     time.Duration
	logger    *slog.Logger
}

var _ rules.Evaluator = AutoAwayRule{}

func LoadAutoAwayRule(cfg configuration.AutoAwayConfiguration, update poller.Update, logger *slog.Logger) (AutoAwayRule, error) {
	usernames := set.New[string]()
	for _, username := range cfg.Users {
		if _, ok := update.GetMobileDevice(username); !ok {
			return AutoAwayRule{}, fmt.Errorf("invalid username: %s", username)
		}
		usernames.Add(username)
	}
	return AutoAwayRule{
		usernames: usernames,
		delay:     cfg.Delay,
		logger:    logger.With("rule", "autoAway"),
	}, nil
}

func (a AutoAwayRule) Evaluate(update poller.Update) (action.Action, error) {
	var users, away int
	homeUsers := make([]string, 0, len(a.usernames))
	awayUsers := make([]string, 0, len(a.usernames))
	for device := range update.MobileDevices.GeoTrackedDevices() {
		if a.usernames.Contains(*device.Name) {
			users++
			if *device.Location.AtHome {
				homeUsers = append(homeUsers, *device.Name)
			} else {
				away++
				awayUsers = append(awayUsers, *device.Name)
			}
		}
	}

	var result action.Action
	if users == away {
		result.Reason = makeReason(awayUsers, "away")
		if *update.HomeState.Presence != tado.AWAY {
			result.Delay = a.delay
			result.State = State{mode: action.HomeInAwayMode, homeId: *update.HomeBase.Id}
		}
	} else {
		result.Reason = makeReason(homeUsers, "home")
		if *update.HomeState.Presence != tado.HOME {
			result.State = State{mode: action.HomeInHomeMode, homeId: *update.HomeBase.Id}
		}
	}

	a.logger.Debug("evaluated",
		slog.String("home", string(*update.Presence)),
		slog.Any("devices", update.MobileDevices),
		slog.Any("result", result),
	)

	return result, nil
}

func makeReason(users []string, state string) string {
	var verb string
	if len(users) == 1 {
		verb = "is"
	} else {
		verb = "are"
	}
	return strings.Join(users, ", ") + " " + verb + " " + state
}

File: internal/controller/rules/zonerules/autoaway.lua
function Evaluate(_, state, devices, _)
    if #devices == 0 then
        return state, 0, "no devices found"
    end
    local homeUsers = FilterDevices(devices, true)
    local wantHeating = #homeUsers > 0
    local reason = "all users are away: " .. ListDevices(devices)
    local delay = 900
    if wantHeating then
        reason = "one or more users are home: " .. ListDevices(homeUsers)
        delay = 0
    end
    if state.Heating == wantHeating then
        return state, 0, reason
    end
    return { Overlay = not wantHeating, Heating = wantHeating }, delay, reason
end

function FilterDevices(devices, state)
    local filtered = {}
    for _, device in ipairs(devices) do
        if device.Home == state then
            table.insert(filtered, device)
        end
    end
    return filtered
end

function ListDevices(devices)
    local result = ""
    for _, obj in ipairs(devices) do
        result = result .. obj.Name .. ", "
    end
    return result:sub(1, -3)
end
